nsqd: add support for configurable max-defer-delay for DPUB

c3llus opened the pull request:
## Background
https://github.com/nsqio/nsq/issues/1499

## Changes Done

1. new `max-defer-delay` command line options
2. default `1h` for `max-defer-delay`
3. refer to the new `max-defer-delay` opt instead of the old `max-req-timeout` for:
a. HTTP
b. TCP

## TODO
1. update the docs

c3llus commented:
Hey @mreiferson, just following up on this PR; would really appreciate your review when you have time!

Thanks!

c3llus commented:
hi @mreiferson, just following up on this PR; would really appreciate your review when you have time!

Thanks!

mreiferson commented:
The implementation is fine in general, but a few things:

1. Let's not rename "timeout" to "delay"
2. For users who are currently setting `--max-req-timeout` and expecting it to effect defers, there's probably some backward compatibility we want to add here. Maybe, if `--max-defer-timeout` is not set and `--max-req-timeout` is, we set the defer timeout to the req timeout?

c3llus commented:
Thanks for the feedback, @mreiferson! :)

I’ve updated the naming to stick with timeout instead of delay. Also added backward compatibility, when `--max-defer-timeout` isn’t explicitly set, it falls back to `--max-req-timeout` like before.

Let me know if there’s anything else you’d like to see adjusted!

mreiferson commented:
This is great, thank you! Can you squash down the commits, please?

c3llus commented:
pushed the squashed commits. let me know if you'd like any further tweaks, @mreiferson! :)
